### ISSUE TITLE
Liquid exception finding multiple files even though there's only one

### ISSUE THREAD

**Issue by mtlynch** [mtlynch/mtlynch.io#310]
I'm seeing what seems like it might be a race condition with the liquid exception handling for file includes:

https://gist.github.com/mtlynch/2de09ceeaf1c55193f33c81dcfe62e55

Every few saves, it gets into a state where it claims that multiple includes have the same title, but the titles seem to be configured correctly.

The errors are also accompanied by lines like this:

```
/usr/local/lib/ruby/gems/2.4.0/gems/jekyll-3.8.2/lib/jekyll/reader.rb:42: warning: conflicting chdir during another chdir block
```

which makes me think there's maybe a race condition (although I've been seeing Jekyll produce that warning, seemingly at random, for many months now).

This is on Ubuntu 16.

```
$ ruby --version
ruby 2.4.4p296 (2018-03-28 revision 63013) [x86_64-linux]
```

**Comment by mtlynch** [comment]
I created a docker config so that it's easy to reproduce this:

https://github.com/mtlynch/mtlynch.io/tree/exception-repro

Run:

```bash
docker-compose up
```

And try opening http://localhost:4000/admin/. If the error doesn't come up immediately, try tweaking files in _posts to see if it triggers.

**Comment by shredtechular** [comment]
Still looking into this, but wanted to provide my findings so far:

- Was able to duplicate the issue in the docker container provided, as well as locally outside the docker container
- Can only replicate the issue when using Jekyll-admin.  Editing files outside of Jekyll-admin, I can't reproduce.  So, I agree this is a race condition and I think it has to do with Jekyll-admin specifically.  It's looping through collections twice or at random to me, which makes no sense...  In the Jekyll-admin repo, I see some past race condition issues that were "fixed" that look awfully similar to the `conflicting chdir during another chdir block` errors.  Also, I am getting `/jekyll-admin-0.8.1/lib/jekyll-admin/server/collection.rb:32:in `block (2 levels) in <class:Server>'` errors outside the docker container
- wondering if Ruby version has something to do with these issues?
- I think I can reproduce the race condition without the exception plugin installed.  After changing the exceptions to warnings, when I look in the browser it is displaying multiple files in the web page. (needs more testing...)
- Options that I'm considering so far:
1. can we do file includes without collections?
2. do we need to find another solution for the exception handling?
3. is there anyway I can dig into the Jekyll-admin code to find the race condition?

**Comment by mtlynch** [comment]
Thanks for the updates!

It sounds like the root cause might be in jekyll-admin. Can you try to generate a simple repro to file a bug on that project? If possible, just in a bare Jekyll project so there's less complexity, but if that's hard to do, maybe just point them to a repro using the mtlynch.io code.

**Comment by shredtechular** [comment]
That's a great idea to generate a very basic jekyll project and see if the issue can be replicated in the most simple of projects.  I will get that done this week and report back.

In the meantime, I do have a fully tested workaround solution ready to go that will check to see if the collection is looping through more than once and if so, ignore subsequent looping.  This is working well and actually will also guard against displaying multiple double file includes that it is also doing currently when using Jekyll-admin.  I'm going to hold off on submitting this PR though until I can get the basic project up and see if the issue exists in that basic project.  If I can't replicate the issue in the basic project, then I'm concerned there is something else going on, so just want to have a little more confidence that the issue is in the Jekyll-admin code vs. our code.

**Comment by mtlynch** [comment]
Can you take another pass at finding the root cause? If it's something in jekyll-admin, I'd like us to be able to submit a bug to them, potentially even a fix.

If it's more than a couple hours and we can't find a root cause, then we can go ahead with the workaround.

**Comment by shredtechular** [comment]
To provide an update and further document some of my findings on this issue:

1. Put Jekyll-admin into the most basic site I can create with a few posts and collections to see if race condition can be replicated.  I could **NOT** replicate the issue in this base site.

2. Put Jekyll-admin into the minimal-mistakes base theme along with several other posts to see if race condition existed in base-theme and I was **NOT** able to reproduce the race condition.  So, I think we can eliminate anything with the parent theme causing the issue.

3. I then went back to the very first commit of mtlynch again to see if I could replicate.  I believe in my previous tests, I didn't update Jekyll admin to the latest version of the Gem when I went to the initial commit of mtlynch, so wanted to revert back but still use the latest Gemfile with the original commit.  I **was still** seeing the race condition with this configuration however.

4. So, next I started trying several different things like changing settings in _config, removing collections, non-critical build functionality, etc.  Still seeing race conditions.

5. Finally, I removed the directories in `/images/` that started with a date similar to how a _post might have been titled and I no longer was able to reproduce the race condition.  I'm not sure if this was the actual cause or if this somehow changed the build times so that the race condition is no longer occurring?  Is the cause the image directory filenames in this version of the site?  In the current version of the site, we no longer name image directories like this, so there still seems to be something else I'm overlooking - perhaps there is an issue with the jekyll-image-responsive gem and jekyll-admin?  I made a temporary repo here of this codebase where I can no longer reproduce the race condition: https://github.com/shredtechular/mtlynch-no-race

**Comment by mtlynch** [comment]
Thanks for the in-depth investigation!

Hmm, I really do wish we could chase down the root cause, but I feel like it's not worth the time, especially if it doesn't seem to be coming from minimal-mistakes or jekyll-admin directly. If we find out the bug is from combining jekyll-image-responsive and jekyll-admin, the maintainers might not want to fix it, so it would be even more of a time sink to fix it ourselves.

I think we should go ahead with the workaround.

**Comment by shredtechular** [comment]
I agree with your logic and thought process there.  I am also very curious about the root cause still, but I can continue to try and find a root cause as time allows and not hold up other issues doing so.

Just as a point of clarity that I tried to express [here](https://github.com/mtlynch/mtlynch.io/pull/316#issuecomment-418202169), is that I'm not sure "workaround" was a great choice of words.  I do feel as though the "workaround" is really a more robust solution regardless if root cause was found or not.  In other words, even if we fixed the root cause, I would recommend implementing the [updated solution](https://github.com/mtlynch/mtlynch.io/pull/316) for checking file includes.

At any rate, in order to implement the "workaround" are the [steps in your last comment](https://github.com/mtlynch/mtlynch.io/pull/316#pullrequestreview-155765015) still the game plan?  Or would you prefer we scrap #310 entirely and I can re-submit a whole new PR that includes your refactored code along with the new robust solution captured in #310?

**Comment by mtlynch** [comment]
Oh right, I forgot that the workaround was still gated on my PR. 

>At any rate, in order to implement the "workaround" are the steps in your last comment still the game plan? 

Yep, that sounds good.

Okay, I rebased and created a new PR. Can you review?

https://github.com/mtlynch/mtlynch.io/pull/355
